### PR TITLE
Feat: Blueprint config restructure and cache

### PR DESCRIPTION
**Pull request by ianshade**
* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
This PR restructures blueprint config stored by Studios and ShowStyles, adds caching of the config, and adds support for a blueprint callback used to process the config before storing it in cache.

* **What is the current behavior?** (You can also link to an open issue here)

1. Studio, ShowStyleBase and ShowStyleVariant store their blueprint-related config in `config` property, as an array.
2. Obtaining the configs always requires database requests and repeating the same processing in blueprints, including populating the config with the default values.

* **What is the new behavior (if this is a feature change)?**

1. Studio, ShowStyleBase and ShowStyleVariant store their blueprint-related config in `blueprintConfig` property, as an object.
2. Core caches the configs in memory, and invalidates the cache only when a config gets modified or when the rundown gets activated.
3. Config can be processed by the blueprints once and stored in the processed form in the cache on Core's request. Optional `parseConfig` blueprint callbacks that were added to `StudioBlueprintManifest` and `ShowStyleBlueprintManifest`, will be called when there is no valid config in cache. Core will cache the processed configs and return them with context's `getStudioConfig` and `getShowStyleConfig` methods. If `parseConfig` is not provided, context will return the configs in an unprocessed form.
4. The default config values are populated by Core.

* **Other information**:
Depends on https://github.com/nrkno/tv-automation-sofie-blueprints-integration/pull/73

**Status**

- [ ] The functionality has been tested by the PR author
- [ ] The functionality has been tested by NRK

### CODE REVIEW

File: meteor/lib/collections/ShowStyleVariants.ts [hunk 1]
```diff
@@ -36,20 +36,13 @@ export function createShowStyleCompound(
...
 
-	let configs: { [id: string]: IConfigItem } = {}
-	_.each(showStyleBase.config, (config: IConfigItem) => {
-		configs[config._id] = config
-	})
-	// Override base configs with variant configs:
-	_.each(showStyleVariant.config, (config: IConfigItem) => {
-		configs[config._id] = config
-	})
+	let configs = { ...showStyleBase.blueprintConfig, ...showStyleVariant.blueprintConfig }
```
[the remark this answers is not in the document]
Comment (ianshade) [reply meteor/lib/collections/ShowStyleVariants.ts]: I added `deepmerge` library for that, and expanded a test case to verify correctness of merging.

File: meteor/client/ui/SegmentTimeline/SegmentTimelineContainer.tsx [hunk 2]
```diff
@@ -205,8 +205,8 @@ export const SegmentTimelineContainer = translateWithTracker<IProps, IState, ITr
 		if (
 			typeof props.studio !== typeof nextProps.studio ||
 			!_.isEqual(props.studio.settings, nextProps.studio.settings) ||
-			!_.isEqual(props.studio.config, nextProps.studio.config) ||
-			!_.isEqual(props.showStyleBase.config, nextProps.showStyleBase.config) ||
+			!_.isEqual(props.studio.blueprintConfig, nextProps.studio.blueprintConfig) ||
```
[the remark this answers is not in the document]
Comment (ianshade) [reply meteor/client/ui/SegmentTimeline/SegmentTimelineContainer.tsx]: Removed.

File: meteor/server/api/blueprints/context/context.ts [hunk 3]
```diff
@@ -155,6 +163,12 @@ export class NotesContext extends CommonContext implements INotesContext {
 	}
 }
 
+const studioBlueprintConfigCache: { [id: string]: Cache } = {}
+const showStyleBlueprintConfigCache: { [id: string]: { [id: string]: Cache } } = {}
+interface Cache {
+	config: any
```
[the remark this answers is not in the document]
Comment (ianshade) [reply meteor/server/api/blueprints/context/context.ts]: Changed to `unknown` in the cache, and in `getStudioConfig` and `getShowStyleConfig` return types. The blueprints-integration PR also has the returned types updated.

**Comment by nytamin** [comment]
I believe that as of the ActivationCache #305 we can re-work this into something that uses that instead of the manual preparations in blueprints that this PR is aiming to do. I'm curious to see if we can reach similar results with that approach, while having keeping the blueprint-integration changes to a minimum.
@ianshade let's discuss this! :)

**Comment by ianshade** [comment]
In order to avoid calculating the same things every time, we would like to be able to cache some data prepared by the blueprints (e.g. arrays of sources and audio layers which are based on the config), but it could be stored separately from the config to avoid the `unknown` type. This PR also changes the structure of the config, changes where the default values are applied etc., so if we want that, I think we can't avoid some changes in blueprints-integration.

File: meteor/server/migration/1_11_0.ts [hunk 4]
```diff
@@ -34,6 +38,9 @@ addMigrationSteps('1.11.0', [
 			}
 		},
 	},
+	migrateConfigToBlueprintConfig('Migrate config to blueprintConfig in Studios', Studios),
```
[the remark this answers is not in the document]
Comment (ianshade) [reply meteor/server/migration/1_11_0.ts]: Moved to X_X_X.

File: meteor/server/migration/lib.ts [hunk 5]
```diff
@@ -251,3 +252,33 @@ export function renamePropertiesInCollection<T extends DBInterface, DBInterface
 		},
 	}
 }
+
+export function migrateConfigToBlueprintConfig<
```
[the remark this answers is not in the document]
Comment (ianshade) [reply meteor/server/migration/lib.ts]: Moved to X_X_X.

**Review by Julusian — APPROVED** [review]

**Review by nytamin — APPROVED** [review]
Looks good!

**Pull request merged**